src/shelly_hap_rgb.cpp

```diff
@@ -0,0 +1,480 @@
+  // Validation.
+  if (cfg.name != nullptr && strlen(cfg.name) > 64) {
+    return mgos::Errorf(STATUS_INVALID_ARGUMENT, "invalid %s",
+                        "name (too long, max 64)");
+  }
+  if (cfg.in_mode != -2 &&
+      (cfg.in_mode < 0 || cfg.in_mode >= (int) InMode::kMax)) {
+    return mgos::Errorf(STATUS_INVALID_ARGUMENT, "invalid %s", "in_mode");
+  }
+  if (cfg.initial_state < 0 || cfg.initial_state >= (int) InitialState::kMax ||
+      (cfg_->in_mode == -1 &&
+       cfg.initial_state == (int) InitialState::kInput)) {
+    return mgos::Errorf(STATUS_INVALID_ARGUMENT, "invalid %s", "initial_state");
+  }
+  cfg.auto_off = (cfg.auto_off != 0);
+  if (cfg.initial_state < 0 || cfg.initial_state > (int) InitialState::kMax) {
+    return mgos::Errorf(STATUS_INVALID_ARGUMENT, "invalid %s", "initial_state");
+  }
+  // Now copy over.
+  if (cfg_->name != nullptr && strcmp(cfg_->name, cfg.name) != 0) {
+    mgos_conf_set_str(&cfg_->name, cfg.name);
+    *restart_required = true;
+  }
+  if (cfg.in_mode != -2 && cfg_->in_mode != cfg.in_mode) {
+    if (cfg_->in_mode == (int) InMode::kDetached ||
+        cfg.in_mode == (int) InMode::kDetached) {
+      *restart_required = true;
+    }
+    cfg_->in_mode = cfg.in_mode;
+  }
+  if (in_inverted != -1 && cfg_->in_inverted != in_inverted) {
+    cfg_->in_inverted = in_inverted;
+    *restart_required = true;
+  }
+  cfg_->initial_state = cfg.initial_state;
+  cfg_->auto_off = cfg.auto_off;
+  cfg_->auto_off_delay = cfg.auto_off_delay;
+  return Status::OK();
+}
+
+void RGB::SaveState() {
+  if (!dirty_) return;
+  mgos_sys_config_save(&mgos_sys_config, false /* try_once */, NULL /* msg */);
+  dirty_ = false;
+}
+
+Status RGB::SetState(const std::string &state_json) {
+  bool state;
+  int brightness, hue, saturation;
+  json_scanf(state_json.c_str(), state_json.size(),
+            "{state: %B, brightness: %d, hue: %d, saturation: %d}",
+            &state, &brightness, &hue, &saturation);
+
+  if (cfg_->state != state) {
+    cfg_->state = state;
+    dirty_ = true;
+  }
+  if (cfg_->brightness != brightness) {
+    cfg_->brightness = brightness;
+    dirty_ = true;
+  }
+  if (cfg_->hue != hue) {
+    cfg_->hue = hue;
+    dirty_ = true;
+  }
+  if (cfg_->saturation != saturation) {
+    cfg_->saturation = saturation;
+    dirty_ = true;
+  }
+
+  if (dirty_) {
+    SetOutputState("RPC");
+  }
+
+  return Status::OK();
+}
+
+void RGB::AutoOffTimerCB() {
+  // Don't set state if auto off has been disabled during timer run
+  if (!cfg_->auto_off) return;
+  if (static_cast<InMode>(cfg_->in_mode) == InMode::kActivation &&
+      in_ != nullptr && in_->GetState() && cfg_->state) {
+    // Input is active, re-arm.
+    LOG(LL_INFO, ("Input is active, re-arming auto off timer"));
+    auto_off_timer_.Reset(cfg_->auto_off_delay * 1000, 0);
+    return;
+  }
+  cfg_->state = false;
+  SetOutputState("auto_off");
+}
+
+void RGB::InputEventHandler(Input::Event ev, bool state) {
```

Comment:
This is currently redundant to `ShellySwitch::InputEventHandler`, will move that to a separate class before this PR is finished.